bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,479 @@
+
+    fn handle_response(
+        self,
+        common: &mut FsmCommonData,
+        from: Baseboard,
+        request_id: Uuid,
+        response: ResponseType,
+    ) -> (State, Output);
+
+    fn tick(self, common: &mut FsmCommonData) -> (State, Output);
+
+    /// A peer has connected.
+    ///
+    /// Outstanding requests may need to get re-sent.
+    /// Since we are using TCP at the network layer, we rely on connections
+    /// to tell us to retry rather than timeouts.
+    ///
+    /// This call can never trigger a state transition, so we don't bother
+    /// taking `self` by value.
+    fn on_connect(
+        &mut self,
+        common: &mut FsmCommonData,
+        peer: Baseboard,
+    ) -> Output;
+
+    /// A peer has disconnected.
+    ///
+    /// We can use this knowledge to assume a request won't complete. For
+    /// example, we may want to move onto the next peer immediately rather than
+    /// wait for a tick.
+    ///
+    /// This call can never trigger a state transition, so we don't bother
+    /// taking `self` by value.
+    fn on_disconnect(
+        &mut self,
+        common: &mut FsmCommonData,
+        peer: Baseboard,
+    ) -> Output;
+}
+
+/// The state machine for a [`$crate::Peer`]
+///
+/// This FSM assumes a network layer above it that can map peer IDs
+/// ( [`Baseboard`]s) to TCP sockets. When an attempt is made to send a
+/// message to a given peer the network layer will send it over an established
+/// connection if one exists. If there is no connection already to that peer,
+/// but the prefix is present and known, the network layer will attempt to
+/// establish a connection and then transmit the message. If the peer is not
+/// known then it must have just had its network prefix removed. In this case
+/// the message will be dropped and the FSM will be told to remove the peer.
+#[derive(Debug)]
+pub struct Fsm {
+    common: FsmCommonData,
+
+    // Use an option to allow taking and mutating `State` independently of `Fsm`
+    // An invariant of the Fsm is that `state` is always `Some` to API callers.
+    state: Option<State>,
+}
+
+impl Fsm {
+    /// Create a new FSM. This is useful when starting from a persistent state
+    /// on disk.
+    pub fn new(id: Baseboard, config: Config, state: State) -> Fsm {
+        Fsm { common: FsmCommonData::new(id, config), state: Some(state) }
+    }
+
+    /// Create a new FSM in `State::Uninitialized`
+    pub fn new_uninitialized(id: Baseboard, config: Config) -> Fsm {
+        let state = Some(State::Uninitialized(UninitializedState {}));
+        Fsm { common: FsmCommonData::new(id, config), state }
+    }
+
+    pub fn state_name(&self) -> &'static str {
+        self.state.as_ref().unwrap().name()
+    }
+
+    pub fn state(&self) -> &State {
+        self.state.as_ref().unwrap()
+    }
+
+    pub fn common_data(&self) -> &FsmCommonData {
+        &self.common
+    }
+
+    /// This call is triggered locally as a result of RSS running
```

Comment:
Clarify whether this is called on one sled or all sleds; I _think_ it's one sled but it's not obvious.

Reply:
Clarified in [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)